packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -0,0 +1,660 @@
+    // argument if the constructor is known to be an upgradable custom element.
+    const elm = (vnode.elm = new UpgradableConstructor((elm: HTMLElement) => {
+        if (!isUndefined(getAssociatedVMIfPresent(elm))) {
+            // There is a possibility that a custom element is registered under tagName, in which
+            // case, the initialization is already carry on, and there is nothing else to do here
+            // since this hook is called right after invoking `document.createElement`.
+            return;
+        }
+
+        const { mode, ctor } = vnode;
+        const def = getComponentInternalDef(ctor);
+
+        vm = vnode.vm = createVM(elm, def, {
+            mode,
+            owner,
+            tagName: sel,
+            renderer,
+        });
+    }));
+
+    linkNodeToShadow(elm, owner);
+
+    if (!isUndefined(vm)) {
+        allocateCustomElementChildren(vnode, vm);
+    } else if (vnode.ctor !== UpgradableConstructor) {
+        throw new TypeError(`Incorrect Component Constructor`);
+    }
+
+    applyStyleScoping(elm, vnode);
+    patchElementAttrsAndProps(null, vnode);
+
+    insertVNode(vnode, parent, anchor);
+
+    if (!isUndefined(vm)) {
+        if (process.env.NODE_ENV !== 'production') {
+            assert.isTrue(vm.state === VMState.created, `${vm} cannot be recycled.`);
+        }
+        runConnectedCallback(vm);
+    }
+
+    mountChildren(vnode.children, elm);
+
+    if (vm) {
+        appendVM(vm);
+    }
+}
+
+function patchCustomElement(n1: VCustomElement, n2: VCustomElement) {
+    const elm = (n2.elm = n1.elm!);
+    const vm = (n2.vm = n1.vm);
+
+    patchElementAttrsAndProps(n1, n2);
+
+    if (!isUndefined(vm)) {
+        // in fallback mode, the allocation will always set children to
+        // empty and delegate the real allocation to the slot elements
+        allocateCustomElementChildren(n2, vm);
+    }
+
+    // in fallback mode, the children will be always empty, so, nothing
+    // will happen, but in native, it does allocate the light dom
+    patchChildren(elm, n1.children, n2.children);
+
+    if (!isUndefined(vm)) {
+        if (process.env.NODE_ENV !== 'production') {
+            assert.isTrue(
+                isArray(n2.children),
+                `Invalid vnode for a custom element, it must have children defined.`
+            );
+        }
+        // this will probably update the shadowRoot, but only if the vm is in a dirty state
+        // this is important to preserve the top to bottom synchronous rendering phase.
+        rerenderVM(vm);
+    }
+}
+
+function unmount(vnode: VNode, parent: ParentNode) {
+    removeVNode(vnode, parent);
+
+    switch (vnode.type) {
+        case VNodeType.Element:
+            unmountChildren(vnode.children, vnode.elm!);
+            break;
+
+        case VNodeType.CustomElement: {
+            const vm = getAssociatedVMIfPresent(vnode.elm);
```

Comment:
@pmdartus in your original refactor, we were doing `vnode.vm`, and I changed it back to `getAssociatedVMIfPresent`. What get's my attention is that it only was observable on the hydration tests. I can say for sure out the top of my mind, but maybe we should set a VM in the vnode as part of the hydration process?